nat/service_ics.go

```diff
@@ -0,0 +1,147 @@
+		return err
+	}
+	if _, err := powerShell("Start-Service -Name RemoteAccess"); err != nil {
+		return err
+	}
+
+	ifaceName, err := getPublicInterfaceName()
+	if err != nil {
+		return err
+	}
+
+	// Enable internet connection sharing for the public interface.
+	_, err = powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.EnableSharing(0)`)
+
+	return err
+}
+
+func (nat *serviceICS) Add(rule RuleForwarding) error {
+	// Enable internet connection sharing for the local interface.
+	// TODO detect interface name from the rule.
+	// TODO firewall rule configuration should be added here for new connections.
+	ifaceName := "myst1"
+
+	_, err := powerShell(`regsvr32 /s hnetcfg.dll;
+	$netShare = New-Object -ComObject HNetCfg.HNetShare;
+	$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+	$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+	$config.EnableSharing(1)`)
+
+	return err
+}
+
+func (nat *serviceICS) Del(rule RuleForwarding) error {
+	// Disable internet connection sharing for the local interface.
+	// TODO detect interace name from the rule.
+	// TODO firewall rule configuration should be added here for cleaning up unused connections.
+	ifaceName := "myst1"
+
+	_, err := powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.DisableSharing()`)
+
+	return err
+}
+
+func (nat *serviceICS) Disable() error {
+	// TODO stop internet connection sharing on all enabled interfaces (nat.ifaces).
+	// TODO we should cleanup as much as possible, not failing after errors.
+	_, err := powerShell("Set-Service -Name RemoteAccess -StartupType " + nat.remoteAccessStatus)
+	if err != nil {
+		return err
+	}
+
+	ifaceName, err := getPublicInterfaceName()
+	if err != nil {
+		return err
+	}
+
+	// Disable internet connection sharing for the public interface.
+	_, err = powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.DisableSharing()`)
+
+	return err
+}
+
+func getPublicInterfaceName() (string, error) {
+	out, err := powerShell(`Get-WmiObject -Class Win32_IP4RouteTable | where { $_.destination -eq '0.0.0.0' -and $_.mask -eq '0.0.0.0'} | foreach { $_.InterfaceIndex }`)
```

Comment:
Have you tested this with win10 home only? I suspect we might have trouble executing those on other versions. We probably need  a way to identify win version and execute appropriately or test on all win7+ versions.

Reply:
Yes, tested on Windows 10 only.
Isn't Windows 7 too old to support it?

Reply:
google "windows versions market share", almost 34% atm ;)

Reply:
Tested it on Windows 7, it works with no issue.